product_tax_multicompany_default/models/product.py

```diff
@@ -0,0 +1,257 @@
+        for k in list(taxes_bd.keys()):
+            if k not in all_companies.ids or k == current_company.id:
+                taxes_bd.pop(k)
+        if tax_type == "taxes":
+            current_tax = current_product.taxes_id
+            field_name_data = "account_sale_tax_id"
+        elif tax_type == "purchase":
+            current_tax = current_product.supplier_taxes_id
+            field_name_data = "account_purchase_tax_id"
+        propagate_taxes = False
+
+        # Tax-free
+        if not current_tax:
+            # Since all companies can have an empty tax field,if there is data
+            # in the database, it means that we can propagate leaving it empty.
+            if len(taxes_bd) > 0:
+                propagate_taxes = True
+        # No tax in other products
+        elif not taxes_bd:
+            # If we do not have any taxes in the database and we add a tax, it means
+            # that we can update the product taxes.
+            current_tax_ids = current_tax.filtered(
+                lambda t: t.company_id == current_company
+            ).ids
+            candidates = set()
+            all_companies = all_companies.filtered(lambda c: c.id != current_company.id)
+            for company in all_companies:
+                candidates.update(
+                    self.taxes_by_company(field_name_data, company, current_tax_ids)
+                )
+            if len(candidates) > 0:
+                propagate_taxes = True
+        else:
+            current_names = current_tax.filtered(
+                lambda t: t.company_id == current_company
+            ).mapped("name")
+            current_tax_ids = current_tax.filtered(
+                lambda t: t.company_id == current_company
+            ).ids
+            for company_id, value in taxes_bd.items():
+                if current_names == list(value.values()):
+                    continue
+                # We are looking to see if there are any taxes that can be applied to
+                # other companies from the taxes that the current product has.
+                candidates = self.taxes_by_company(
+                    field_name_data,
+                    self.env["res.company"].browse(company_id),
+                    current_tax_ids,
+                )
+                if len(candidates) == len(current_tax):
+                    propagate_taxes = True
+                    break
+
+        return propagate_taxes
+
+    def _get_product_taxes(self, field, product_id):
+        """Returns taxes on purchase or sales taxes
+
+        We need the product taxes for other companies, and we cannot take
+        them from the object itself, so we have to consult the database
+        where the product taxes passed by parameter are located for all
+        companies.
+
+        :param field: can be 'taxes' or 'purchase'
+        :param product_id: product to which the query should be applied
+        """
```

Comment:
Make it available only for one product

```suggestion
    def _get_product_taxes(self, field):
        """Returns taxes on purchase or sales taxes

        We need the product taxes for other companies, and we cannot take
        them from the object itself, so we have to consult the database
        where the product taxes passed by parameter are located for all
        companies.

        :param field: can be 'taxes' or 'purchase'
        """
        self.ensure_one()
```